analysis/pseudobulk-bulk-prediction/notebooks/perform-ora.Rmd

```diff
@@ -0,0 +1,270 @@
+  purrr::map(\(i) {
+    
+    shuffled_df <- data.frame(
+      ensembl_id = summary_residual_df$ensembl_id, 
+      summary_residual = sample(summary_residual_df$summary_residual, nrow(summary_residual_df))   
+    )
+    
+    geneset_list |>
+      purrr::map(
+        \(gs_df) analyze_outliers(shuffled_df, gs_df$ensembl_id)
+      ) |>
+      purrr::list_rbind(names_to = "geneset_cell_type") |>
+      # rename columns to with shuffled_ prefix for code clarity
+      dplyr::rename(
+        shuffled_or_bulk = or_bulk, 
+        shuffled_or_pseudobulk = or_pseudobulk
+      ) 
+  }) |> purrr::list_rbind(names_to = "replicate")
+```
+
+
+Below, we visualize the shuffled odds ratio distributions, with a red line representing the odds ratio calculated from the real data.
+
+```{r fig.height=10, fig.width=10}
+ggplot(shuffled_or_df) + 
+  aes(x = shuffled_or_bulk) + 
+  geom_histogram(binwidth = 0.2, fill = "gray65") +
+  facet_wrap(vars(geneset_cell_type), scales = "free_x") +
+  geom_vline(
+    data = baseline_or_df, 
+    aes(xintercept = or_bulk), 
+    color = "firebrick", 
+    linewidth = 0.75
+  ) + 
+  labs(
+    title = "Odds ratio distributions in the bulk direction", 
+    x = "Shuffled bulk-direction odds ratios"
+  )
+```
+
+
+```{r fig.height=10, fig.width=10}
+ggplot(shuffled_or_df) + 
+  aes(x = shuffled_or_pseudobulk) + 
+  geom_histogram(binwidth = 0.2, fill = "gray65") +
+  facet_wrap(vars(geneset_cell_type), scales = "free_x") +
+  geom_vline(
+    data = baseline_or_df, 
+    aes(xintercept = or_pseudobulk), 
+    color = "firebrick", 
+    linewidth = 0.75
+  ) + 
+  labs(
+    title = "Odds ratio distributions in the pseudobulk direction", 
+    x = "Shuffled pseudobulk-direction odds ratios"
+  )
+```
+
+
+Next, we can calculate P-values for each gene set.
+We'll also multiply each final P-value by 2 since we are performing 2 tests per cell type.
+
+```{r}
+permutation_pvalues <- geneset_names |>
+  purrr::set_names(geneset_names) |>
+  purrr::map(
+    \(geneset_name) {
+      
+    # This will have one row for the given geneset
+    geneset_baseline_df <- baseline_or_df |>
+      dplyr::filter(geneset_cell_type == geneset_name)
+    
+    # Define these variables for code clarity
+    baseline_bulk_or <- geneset_baseline_df$or_bulk
+    baseline_pseudobulk_or <- geneset_baseline_df$or_pseudobulk
+      
+    shuffled_or_df |>
+      dplyr::filter(geneset_cell_type == geneset_name) |>
+      # Calculate P-value as fraction of shuffled ORs >= baseline OR
+      dplyr::summarize(
+        bulk_pvalue = sum(shuffled_or_bulk >= baseline_bulk_or) / params$reps, 
+        pseudobulk_pvalue = sum(shuffled_or_pseudobulk >= baseline_pseudobulk_or) / params$reps
+      ) |>
+      # correct pvalue for 2 tests, capping it at 1
+      dplyr::mutate(
+        bulk_pvalue = min(2*bulk_pvalue, 1),
+        pseudobulk_pvalue = min(2*pseudobulk_pvalue, 1),
+      )
+    }) |>
+    purrr::list_rbind(names_to = "geneset_cell_type")
```

Comment:
As discussed in Slack, calculate an FDR using the p-values from both "directions."